src/lib/slot.c

```diff
@@ -139,8 +144,23 @@ CK_RV slot_mechanism_list_get (CK_SLOT_ID slot_id, CK_MECHANISM_TYPE *mechanism_
         return CKR_BUFFER_TOO_SMALL;
     }
 
-    *count = ARRAY_LEN(mechs);
-    memcpy(mechanism_list, mechs, sizeof(mechs));
+    TPMS_CAPABILITY_DATA *capabilityData = NULL;
+    rv = tpm_get_algorithms (t->tctx, &capabilityData);
+    if (rv != CKR_OK) {
+        LOGE("Retrieving supported algorithms from TPM failed");
+        return rv;
+    }
+    TPMU_CAPABILITIES *algs= &capabilityData->data;
+
+    for (unsigned int i = 0; i < ARRAY_LEN(mechs); i++) {
+        if (tpm_algs_is_mechanism_supported(algs, mechs[i])) {
```

Comment:
Im confused by this interface, hey TPM what algorithms you got, are they supported? I would imagine the interface to the tpm layer should just be, give me a list of pkcs11 mechs you support. Rather than this we have a list, now filter.

Reply:
The list is what the library supports in general, which is then filtered by the TPM capabilities.
That's the rationale behind this idea.
We can hide it, okay.